In [26]:
import pandas as pd
import json

In [36]:
# Load data from the JSON file
with open('optimization_results.json', 'r') as json_file:
    data = json.load(json_file)

In [55]:
# Create a Pandas DataFrame
df2 = pd.DataFrame(data)

cols = ['TL', 'imgsz', 'epochs','Precision', 'Recall',
       'mAP@.5', 'mAP@.5:.95','optimization_technique', 'model_performance', ]

df2[cols]

TL  imgsz  epochs  Precision  Recall  mAP@.5  mAP@.5:.95  \
0  yolov8s    800     100      0.927   0.867   0.925       0.702   
1  yolov8s    800     100      0.924   0.868   0.926       0.691   
2  yolov8m    640     100      0.941   0.873   0.934       0.728   
3  yolov8m    640     100      0.939   0.871   0.932       0.713   
4  yolov8l    640      50      0.921   0.889   0.941       0.723   
5  yolov8l    640      50      0.920   0.883   0.937       0.712   
6  yolov8m    640      50      0.936   0.860   0.932       0.721   
7  yolov8m    640      50      0.939   0.854   0.933       0.710   
8  yolov8s    800      60      0.922   0.873   0.931       0.704   
9  yolov8s    800      60      0.927   0.873   0.929       0.690   

  optimization_technique  model_performance  
0         Floating-point               4.85  
1              Quantized               8.35  
2         Floating-point               1.79  
3              Quantized               3.18  
4         Floating-point               0.80  
5              Quantized               1.54  
6         Floating-point               1.76  
7              Quantized               3.14  
8         Floating-point               4.76  
9              Quantized               8.18

In [59]:
OUTPUT_DIR = "../003-Evaluating-models/output_csvs"
        
df = pd.read_csv(f"{OUTPUT_DIR}/eval_trained_on_dataset_v2_task_val_ds_original.csv")

columns_of_interest = ['transfer_learning', 'imgsz','epochs','all_P', 'all_R','all_mAP@.5', 'all_mAP@.5:.95']

df.rename(columns=dict(zip(columns_of_interest, cols)), inplace=True)

df1 = df.loc[[2,3,1,7,9]]
df1

model_key   model       TL  imgsz  \
2  model_yolov8_tf_yolov8m_imgsz_640_epochs_100_b...  yolov8  yolov8m    640   
3  model_yolov8_tf_yolov8m_imgsz_640_epochs_50_ba...  yolov8  yolov8m    640   
1  model_yolov8_tf_yolov8l_imgsz_640_epochs_50_ba...  yolov8  yolov8l    640   
7  model_yolov8_tf_yolov8s_imgsz_800_epochs_100_b...  yolov8  yolov8s    800   
9  model_yolov8_tf_yolov8s_imgsz_800_epochs_60_ba...  yolov8  yolov8s    800   

   epochs  batch dataset optimizer   lr0  Precision  Recall    all_F1  mAP@.5  \
2     100     16      v2       SGD  0.01      0.926   0.891  0.454081   0.941   
3      50     16      v2       SGD  0.01      0.925   0.874  0.449389   0.938   
1      50     16      v2       SGD  0.01      0.927   0.880  0.451444   0.944   
7     100     16      v2       SGD  0.01      0.922   0.883  0.451039   0.938   
9      60     16      v2       SGD  0.01      0.931   0.879  0.452127   0.940   

   mAP@.5:.95  
2       0.743  
3       0.731  
1       0.733  
7       0.728  
9       0.723

In [70]:
pd.concat([df1, df2])[cols].sort_values(by=['TL', 'imgsz', 'epochs','Recall'])
pd.concat([df1, df2])[cols].sort_values(by=['Precision'])

TL  imgsz  epochs  Precision  Recall  mAP@.5  mAP@.5:.95  \
5  yolov8l    640      50      0.920   0.883   0.937       0.712   
4  yolov8l    640      50      0.921   0.889   0.941       0.723   
7  yolov8s    800     100      0.922   0.883   0.938       0.728   
8  yolov8s    800      60      0.922   0.873   0.931       0.704   
1  yolov8s    800     100      0.924   0.868   0.926       0.691   
3  yolov8m    640      50      0.925   0.874   0.938       0.731   
2  yolov8m    640     100      0.926   0.891   0.941       0.743   
1  yolov8l    640      50      0.927   0.880   0.944       0.733   
0  yolov8s    800     100      0.927   0.867   0.925       0.702   
9  yolov8s    800      60      0.927   0.873   0.929       0.690   
9  yolov8s    800      60      0.931   0.879   0.940       0.723   
6  yolov8m    640      50      0.936   0.860   0.932       0.721   
3  yolov8m    640     100      0.939   0.871   0.932       0.713   
7  yolov8m    640      50      0.939   0.854   0.933       0.710   
2  yolov8m    640     100      0.941   0.873   0.934       0.728   

  optimization_technique  model_performance  
5              Quantized               1.54  
4         Floating-point               0.80  
7                    NaN                NaN  
8         Floating-point               4.76  
1              Quantized               8.35  
3                    NaN                NaN  
2                    NaN                NaN  
1                    NaN                NaN  
0         Floating-point               4.85  
9              Quantized               8.18  
9                    NaN                NaN  
6         Floating-point               1.76  
3              Quantized               3.18  
7              Quantized               3.14  
2         Floating-point               1.79


The model with the highest performance in terms of Frames Per Second (FPS) is the 'yolov8s' with Quantized optimization technique and 800 imgsz (8.35 FPS).
The lowest performance is seen in the 'yolov8l' model with Floating-point optimization and 800 imgsz (0.8 FPS).


Both Floating-point and Quantized optimization techniques are used.
Quantized models tend to have higher FPS compared to Floating-point models.

The 'yolov8m' model without optimization, trained on 100 epochs and 640 imgsz has the highest recall (0.891).
The 'yolov8m' model with Floating-point optimization, trained on 100 epochs and 640 imgsz has the highest precision (0.941).